projects/packages/autoloader/src/class-latest-autoloader-guard.php

```diff
@@ -68,7 +71,11 @@ public function should_stop_init( $current_plugin, $plugins, $was_included_by_au
 		}
 
 		$latest_plugin = $this->autoloader_locator->find_latest_autoloader( $plugins, $jetpack_autoloader_latest_version );
-		if ( isset( $latest_plugin ) && $latest_plugin !== $current_plugin ) {
+
+		$current_plugin_data = $current_plugin ? get_plugin_data( $current_plugin . '/' . basename( $current_plugin ) . '.php' ) : null;
+		$latest_plugin_data  = $latest_plugin ? get_plugin_data( $latest_plugin . '/' . basename( $latest_plugin ) . '.php' ) : null;
```

Comment:
This code makes an unwarranted assumption the plugin file is named like "fooplugin/fooplugin.php", which is not always the case. For example, testing this using the Jetpack Beta Tester plugin to install Jetpack results in

>  Warning: file_get_contents(.../wp-content/plugins/jetpack-dev/jetpack-dev.php): failed to open stream: No such file or directory in .../wp-includes/functions.php on line 6633

Reply:
Thanks for taking the time to review this. You have raised very good points. I will investigate what can be done to check the plugin files for all scenarios adding as less overhead as possible.

Reply:
@darssen I think you can get inspiration from the way we did it in [the Sync module related to plugins](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/sync/src/modules/class-plugins.php#L363). We have overlooked the necessity of checking whether the file actually exists.

@anomiex I agree that the overhead raises concerns, but do you think it would really be that concerning? The method in question is only run when a plugin tries to initialize an instance of a Jetpack Autoloader. Surely two additional file reads per plugin won't be that much of an issue, or am I missing something?

As to the unwarranted assumption - this is sort of an edge case that we are trying to accomodate. I would suggest against complicating the logic here and assuming that if a user has any other sort of setup, they are on their own. Even this edge case is artificially introduced by running the same version of Jetpack in `mu-plugins`, as well as `plugins`. We haven't seen it in the wild, we're doing it for a certain consumer test scenario.

Reply:
It seems like a hot path, this code is run for every request. So it does seem worthwhile to consider.

Reply:
@zinigor  and I have been discussing that maybe we should not be considering this fix for an edge case since it might impact performance. But prior to definitely discarding it, I wanted to give it another go.
Things I discovered are:
- The main reason why this is a problem is the way we are identifying plugins in the whole autoloader. We consider the folder of the plugin as the identifier and we compared based on that. (That is why I tried comparing by file name, doing a naive assumption.)
- I also realized that the reason why it fatals is that it tries to include "same" autoloader from a different file. This makes the require_once and its internal [get_included_files](https://www.php.net/manual/en/function.get-included-files.php) check unusable for this scenario.
- We could compare based on one of the files of the autoloader (The one that we require in the autoloader guard is a good option).
- The comparison doesn't need to be the whole file, something that checks the beginning of the file would actually be good enough, since each autoloader has a different namespace unless we are in the scenario we are trying to avoid :)
So I have this diff with an initial approach.
```diff
diff --git a/projects/packages/autoloader/src/class-latest-autoloader-guard.php b/projects/packages/autoloader/src/class-latest-autoloader-guard.php
index 54edfc87c2..8cd0e17295 100644
--- a/projects/packages/autoloader/src/class-latest-autoloader-guard.php
+++ b/projects/packages/autoloader/src/class-latest-autoloader-guard.php
@@ -68,11 +68,33 @@ class Latest_Autoloader_Guard {
                }
 
                $latest_plugin = $this->autoloader_locator->find_latest_autoloader( $plugins, $jetpack_autoloader_latest_version );
-               if ( isset( $latest_plugin ) && $latest_plugin !== $current_plugin ) {
+
+               if ( isset( $latest_plugin ) && ! $this->autoloaders_are_similar( $current_plugin, $latest_plugin ) ) {
                        require $this->autoloader_locator->get_autoloader_path( $latest_plugin );
                        return true;
                }
 
                return false;
        }
+
+       /**
+        * Checks if two autoloaders are same by comparing an excerpt of their autoload_packages.php.
+        * Different autoloaders should have different namespace so this should be equivalent to check the whole file.
+        *
+        * @param string $current_plugin The current plugin.
+        * @param string $compare_plugin The plugin to compare the autoloader with.
+        *
+        * @return bool True if autoloaders are similar, otherwise false.
+        */
+       private function are_autoloaders_same( $current_plugin, $compare_plugin ) {
+               $current_autoloader_path = $this->autoloader_locator->get_autoloader_path( $current_plugin );
+               $compare_autoloader_path = $this->autoloader_locator->get_autoloader_path( $compare_plugin );
+
+               // Check if filesize is different
+               if ( filesize( $current_autoloader_path ) !== filesize( $compare_autoloader_path ) ) {
+                       return false;
+               }
+               return file_get_contents( $current_autoloader_path, false, null, 0, 8 * KB_IN_BYTES )
+                       == file_get_contents( $compare_autoloader_path, false, null, 0, 8 * KB_IN_BYTES );
+       }
 }
```
Please @anomiex @zinigor  suggest if this is something I should pursue or not. Or I should not be including any logic into the autoloader which might impact the performance of a hot path for one edge case. (edited)

Reply:
Since the files we're looking at are generated using the composer build process, I think this would be a good way to proceed. This only relies on reading raw file contents, does not need any parsing and can be done much more frequently than parsing plugin meta data.
Do you have any caveats in mind @anomiex ?

Reply:
@darssen - I think it's a great idea to check the `autoload_packages.php` file.